guidance/_parser.py

```diff
@@ -535,12 +535,16 @@ def _record_captures_partial(self, data, log_prob_data):
     def _record_captures_from_root(self, initial_item, data, log_prob_data):
         byte_data = self.bytes
         stack = [(initial_item, 0)]
-        used_names = (
-            set()
-        )  # track which capture names have been used so self-recursive children don't overwrite their parents
-
+        # track which capture names have been used so self-recursive children don't overwrite their parents
+        used_names = set()
+        # track which items we have seen so we don't process them multiple times, leading to infinite loops
+        seen = set()
         while stack:
             item, byte_pos = stack.pop()
+            if (item, byte_pos) in seen:
```

Comment:
Maybe just check `item`... It seems to work fine either way, but I was a bit paranoid since the same rule in a different position doesn't necessarily mean a loop?